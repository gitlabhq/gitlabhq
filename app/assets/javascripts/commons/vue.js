import Vue from 'vue';
import GlLicensedFeaturesPlugin from '~/vue_shared/gl_licensed_features_plugin';
import GlFeatureFlagsPlugin from '~/vue_shared/gl_feature_flags_plugin';
import GlAbilitiesPlugin from '~/vue_shared/gl_abilities_plugin';
import Translate from '~/vue_shared/translate';

if (process.env.NODE_ENV !== 'production') {
  Vue.config.productionTip = false;
}

Vue.use(GlLicensedFeaturesPlugin);

Vue.use(GlFeatureFlagsPlugin);
Vue.use(GlAbilitiesPlugin);
Vue.use(Translate);

Vue.config.ignoredElements = ['gl-emoji'];
