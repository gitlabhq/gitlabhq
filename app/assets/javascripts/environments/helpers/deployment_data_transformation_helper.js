import { getIdFromGraphQLId } from '~/graphql_shared/utils';

/**
 * This function transforms Commit object coming from GraphQL to object compatible with app/assets/javascripts/vue_shared/components/commit.vue author object
 * @param {Object} Commit
 * @returns {Object}
 */
export const getAuthorFromCommit = (commit) => {
  if (commit.author) {
    return {
      username: commit.author.name,
      path: commit.author.webUrl,
      avatar_url: commit.author.avatarUrl,
    };
  }
  return {
    username: commit.authorName,
    path: `mailto:${commit.authorEmail}`,
    avatar_url: commit.authorGravatar,
  };
};

/**
 * This function transforms deploymentNode object coming from GraphQL to object compatible with app/assets/javascripts/vue_shared/components/commit.vue
 * @param {Object} deploymentNode
 * @returns {Object}
 */
export const getCommitFromDeploymentNode = (deploymentNode) => {
  if (!deploymentNode.commit) {
    throw new Error("deploymentNode argument doesn't have 'commit' field", deploymentNode);
  }
  return {
    title: deploymentNode.commit.message,
    commitUrl: deploymentNode.commit.webUrl,
    shortSha: deploymentNode.commit.shortId,
    tag: deploymentNode.tag,
    commitRef: {
      name: deploymentNode.ref,
    },
    author: getAuthorFromCommit(deploymentNode.commit),
  };
};

export const convertJobToDeploymentAction = (job) => {
  return {
    name: job.name,
    playable: job.playable,
    scheduledAt: job.scheduledAt,
    playPath: `${job.webPath}/play`,
  };
};

export const getActionsFromDeploymentNode = (deploymentNode, lastDeploymentName) => {
  if (!deploymentNode || !lastDeploymentName) {
    return [];
  }

  return (
    deploymentNode.job?.deploymentPipeline?.jobs?.nodes
      ?.filter((deployment) => deployment.name !== lastDeploymentName)
      .map(convertJobToDeploymentAction) || []
  );
};

export const getRollbackActionFromDeploymentNode = (deploymentNode, environment) => {
  const { job, id } = deploymentNode;

  if (!job) {
    return null;
  }
  const isLastDeployment = id === environment.lastDeployment?.id;
  const { webPath } = job;
  return {
    id,
    name: environment.name,
    lastDeployment: {
      commit: deploymentNode.commit,
      isLast: isLastDeployment,
    },
    retryUrl: `${webPath}/retry`,
  };
};

const getDeploymentApprovalFromDeploymentNode = (deploymentNode, environment) => {
  if (!environment.protectedEnvironments || environment.protectedEnvironments.nodes.length === 0) {
    return {
      isApprovalActionAvailable: false,
    };
  }

  const protectedEnvironmentInfo = environment.protectedEnvironments.nodes[0];

  const hasApprovalRules = protectedEnvironmentInfo.approvalRules.nodes?.length > 0;
  const hasRequiredApprovals = protectedEnvironmentInfo.requiredApprovalCount > 0;

  const isApprovalActionAvailable = hasRequiredApprovals || hasApprovalRules;
  const requiredMultipleApprovalRulesApprovals =
    protectedEnvironmentInfo.approvalRules.nodes.reduce((requiredApprovals, rule) => {
      return requiredApprovals + rule.requiredApprovals;
    }, 0);

  const requiredApprovalCount = hasRequiredApprovals
    ? protectedEnvironmentInfo.requiredApprovalCount
    : requiredMultipleApprovalRulesApprovals;

  return {
    isApprovalActionAvailable,
    deploymentIid: deploymentNode.iid,
    environment: {
      name: environment.name,
      tier: environment.tier,
      requiredApprovalCount,
    },
  };
};

/**
 * This function transforms deploymentNode object coming from GraphQL to object compatible with app/assets/javascripts/environments/environment_details/page.vue table
 * @param {Object} deploymentNode
 * @returns {Object}
 */
export const convertToDeploymentTableRow = (deploymentNode, environment) => {
  const { lastDeployment } = environment;
  const commit = getCommitFromDeploymentNode(deploymentNode);
  const { job } = deploymentNode;
  const pipeline = job?.downstreamPipeline ? job?.downstreamPipeline : job?.deploymentPipeline;

  return {
    status: deploymentNode.status.toLowerCase(),
    id: deploymentNode.iid,
    triggerer: deploymentNode.triggerer,
    commit,
    job: job && {
      webPath: job.webPath,
      label: `${job.name} (#${getIdFromGraphQLId(job.id)})`,
      pipeline: pipeline?.path && {
        path: pipeline?.path,
        label: `#${getIdFromGraphQLId(pipeline.id)}`,
      },
    },
    created: deploymentNode.createdAt || '',
    finished: deploymentNode.finishedAt || '',
    actions: getActionsFromDeploymentNode(deploymentNode, lastDeployment?.job?.name),
    rollback: getRollbackActionFromDeploymentNode(deploymentNode, environment),
    deploymentApproval: getDeploymentApprovalFromDeploymentNode(deploymentNode, environment),
    webPath: deploymentNode.webPath || '',
  };
};
