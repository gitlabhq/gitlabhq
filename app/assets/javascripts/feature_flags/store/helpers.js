import { ROLLOUT_STRATEGY_GITLAB_USER_LIST, NEW_VERSION_FLAG } from '../constants';

const mapStrategyScopesToRails = (scopes) =>
  scopes.length === 0
    ? [{ environment_scope: '*' }]
    : scopes.map((s) => ({
        id: s.id,
        _destroy: s.shouldBeDestroyed,
        environment_scope: s.environmentScope,
      }));

const mapStrategyScopesToView = (scopes) =>
  scopes.map((s) => ({
    id: s.id,
    // eslint-disable-next-line no-underscore-dangle
    shouldBeDestroyed: Boolean(s._destroy),
    environmentScope: s.environment_scope,
  }));

const mapStrategiesParametersToViewModel = (params) => {
  if (params.userIds) {
    return { ...params, userIds: params.userIds.split(',').join(', ') };
  }
  return params;
};

export const mapStrategiesToViewModel = (strategiesFromRails) =>
  (strategiesFromRails || []).map((s) => ({
    id: s.id,
    name: s.name,
    parameters: mapStrategiesParametersToViewModel(s.parameters),
    userList: s.user_list,
    // eslint-disable-next-line no-underscore-dangle
    shouldBeDestroyed: Boolean(s._destroy),
    scopes: mapStrategyScopesToView(s.scopes),
  }));

const mapStrategiesParametersToRails = (params) => {
  if (params.userIds) {
    return { ...params, userIds: params.userIds.replace(/\s*,\s*/g, ',') };
  }
  return params;
};

const mapStrategyToRails = (strategy) => {
  const mappedStrategy = {
    id: strategy.id,
    name: strategy.name,
    _destroy: strategy.shouldBeDestroyed,
    scopes_attributes: mapStrategyScopesToRails(strategy.scopes || []),
    parameters: mapStrategiesParametersToRails(strategy.parameters),
  };

  if (strategy.name === ROLLOUT_STRATEGY_GITLAB_USER_LIST) {
    mappedStrategy.user_list_id = strategy.userList?.id;
  }
  return mappedStrategy;
};

export const mapStrategiesToRails = (params) => ({
  operations_feature_flag: {
    name: params.name,
    description: params.description,
    active: params.active,
    strategies_attributes: (params.strategies || []).map(mapStrategyToRails),
    version: NEW_VERSION_FLAG,
  },
});
