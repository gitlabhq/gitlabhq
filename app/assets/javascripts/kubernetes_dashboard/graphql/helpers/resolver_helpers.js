import {
  CoreV1Api,
  Configuration,
  WatchApi,
  EVENT_DATA,
  EVENT_TIMEOUT,
  EVENT_ERROR,
} from '@gitlab/cluster-client';
import { getWatchManager } from '~/environments/services/websocket_connection_service';
import { connectionStatus } from '~/environments/graphql/resolvers/kubernetes/constants';
import { updateConnectionStatus } from '~/environments/graphql/resolvers/kubernetes/k8s_connection_status';
import { s__ } from '~/locale';

export const handleClusterError = async (err) => {
  if (!err.response) {
    throw err;
  }

  const contentType = err.response.headers.get('Content-Type');

  if (contentType !== 'application/json') {
    throw new Error(
      s__(
        'KubernetesDashboard|There was a problem fetching cluster information. Refresh the page and try again.',
      ),
    );
  }

  const errorData = await err.response.json();
  throw errorData;
};

export const buildWatchPath = ({ resource, api = 'api/v1', namespace = '' }) => {
  return namespace ? `/${api}/namespaces/${namespace}/${resource}` : `/${api}/${resource}`;
};

export const mapWorkloadItem = (item) => {
  const status = item.status || {};
  const spec = item.spec || {};
  const metadata = {
    ...item.metadata,
    annotations: item.metadata?.annotations || {},
    labels: item.metadata?.labels || {},
  };
  return { status, spec, metadata, __typename: 'LocalWorkloadItem' };
};

export const mapEventItem = ({
  lastTimestamp = '',
  eventTime = '',
  message,
  reason,
  source,
  type,
}) => ({ lastTimestamp, eventTime, message, reason, source, type });

export const subscribeToSocket = async ({ watchId, watchParams, cacheParams, config }) => {
  const { updateQueryCache, updateConnectionStatusFn } = cacheParams;

  try {
    const watcherConnection = getWatchManager(config);

    const watcher = await watcherConnection.initConnection({
      message: { watchId, watchParams },
    });

    const handleConnectionStatus = (status) => {
      if (updateConnectionStatusFn) {
        updateConnectionStatusFn(status);
      }
    };

    watcher.on(EVENT_DATA, watchId, (data) => {
      updateQueryCache(data);
      handleConnectionStatus(connectionStatus.connected);
    });

    watcher.on(EVENT_ERROR, watchId, () => {
      handleConnectionStatus(connectionStatus.disconnected);
    });
  } catch (err) {
    throw new Error(s__('KubernetesDashboard|Failed to establish WebSocket connection'));
  }
};

export const watchWorkloadItems = async ({
  client,
  query,
  configuration,
  namespace,
  watchPath,
  queryField,
  watchParams,
}) => {
  const config = new Configuration(configuration);
  const watcherApi = new WatchApi(config);

  const updateStatus = (status) =>
    updateConnectionStatus(client, {
      configuration,
      namespace,
      resourceType: queryField,
      status,
    });

  const updateQueryCache = (data) => {
    const result = data.map(mapWorkloadItem);
    client.writeQuery({
      query,
      variables: { configuration, namespace },
      data: { [queryField]: result },
    });
  };

  const watchFunction = async () => {
    try {
      const watcher = await watcherApi.subscribeToStream(watchPath, { watch: true });

      watcher.on(EVENT_DATA, (data) => {
        updateQueryCache(data);
        updateStatus(connectionStatus.connected);
      });

      watcher.on(EVENT_TIMEOUT, () => {
        updateStatus(connectionStatus.disconnected);
      });

      watcher.on(EVENT_ERROR, () => {
        updateStatus(connectionStatus.disconnected);
      });
    } catch (err) {
      await handleClusterError(err);
    }
  };

  updateStatus(connectionStatus.connecting);

  const watchId = namespace ? `${queryField}-n-${namespace}` : `${queryField}-all-namespaces`;
  const cacheParams = {
    updateQueryCache,
    updateConnectionStatusFn: updateStatus,
  };

  try {
    await subscribeToSocket({ watchId, watchParams, cacheParams, config });
  } catch {
    await watchFunction();
  }
};

export const getK8sPods = ({
  client,
  query,
  configuration,
  namespace = '',
  mapFn = mapWorkloadItem,
  queryField = 'k8sPods',
}) => {
  const config = new Configuration(configuration);

  const coreV1Api = new CoreV1Api(config);
  const podsApi = namespace
    ? coreV1Api.listCoreV1NamespacedPod({ namespace })
    : coreV1Api.listCoreV1PodForAllNamespaces();

  return podsApi
    .then((res) => {
      const watchPath = buildWatchPath({ resource: 'pods', namespace });
      const watchParams = {
        version: 'v1',
        resource: 'pods',
        namespace,
      };

      watchWorkloadItems({
        client,
        query,
        configuration,
        namespace,
        watchPath,
        queryField,
        watchParams,
      });

      const data = res?.items || [];
      return data.map(mapFn);
    })
    .catch(async (err) => {
      try {
        await handleClusterError(err);
      } catch (error) {
        throw new Error(error.message);
      }
    });
};
