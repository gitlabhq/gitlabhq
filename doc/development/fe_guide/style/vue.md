---
stage: none
group: unassigned
info: Any user with at least the Maintainer role can merge updates to this content. For details, see https://docs.gitlab.com/development/development_processes/#development-guidelines-review.
title: Vue.js style guide
---

## Linting

We default to [eslint-vue-plugin](https://github.com/vuejs/eslint-plugin-vue), with the `plugin:vue/recommended`.
Check the [rules](https://github.com/vuejs/eslint-plugin-vue#bulb-rules) for more documentation.

## Basic Rules

1. Use `.vue` for Vue templates. Do not use `%template` in HAML.
1. Explicitly define data being passed into the Vue app

   ```javascript
   // bad
   return new Vue({
     el: '#element',
     name: 'ComponentNameRoot',
     components: {
       componentName
     },
     provide: {
       ...someDataset
     },
     props: {
       ...anotherDataset
     },
     render: createElement => createElement('component-name'),
   }));

   // good
   const { foobar, barfoo } = someDataset;
   const { foo, bar } = anotherDataset;

   return new Vue({
     el: '#element',
     name: 'ComponentNameRoot',
     components: {
       componentName
     },
     provide: {
       foobar,
       barfoo
     },
     props: {
       foo,
       bar
     },
     render: createElement => createElement('component-name'),
   }));
   ```

   We discourage the use of the spread operator in this specific case in
   order to keep our codebase explicit, discoverable, and searchable.
   This applies in any place where we would benefit from the above, such as
   when [initializing Vuex state](../vuex.md#why-not-just-spread-the-initial-state).
   The pattern above also enables us to easily parse non scalar values during
   instantiation.

   ```javascript
   return new Vue({
     el: '#element',
     name: 'ComponentNameRoot',
     components: {
       componentName
     },
     props: {
       foo,
       bar: parseBoolean(bar)
     },
     render: createElement => createElement('component-name'),
   }));
   ```

## Component usage within templates

1. Prefer a component's kebab-cased name over other styles when using it in a template

   ```html
   // bad
   <MyComponent />

   // good
   <my-component />
   ```

## `<style>` tags

We don't use `<style>` tags in Vue components for a few reasons:

1. You cannot use SCSS variables and mixins or [Tailwind CSS](scss.md#tailwind-css) `@apply` directive.
1. These styles get inserted at runtime.
1. We already have a few other ways to define CSS.

Instead of using a `<style>` tag you should use [Tailwind CSS utility classes](scss.md#tailwind-css) or [page specific CSS](https://gitlab.com/groups/gitlab-org/-/epics/3694).

## Vue testing

Over time, a number of programming patterns and style preferences have emerged in our efforts to
effectively test Vue components. The following guide describes some of these.
**These are not strict guidelines**, but rather a collection of suggestions and good practices that
aim to provide insight into how we write Vue tests at GitLab.

### Mounting a component

Typically, when testing a Vue component, the component should be "re-mounted" in every test block.

To achieve this:

1. Create a mutable `wrapper` variable inside the top-level `describe` block.
1. Mount the component using [`mount`](https://v1.test-utils.vuejs.org/api/#mount) or [`shallowMount`](https://v1.test-utils.vuejs.org/api/#shallowMount).
1. Reassign the resulting [`Wrapper`](https://v1.test-utils.vuejs.org/api/wrapper/#wrapper) instance to our `wrapper` variable.

Creating a global, mutable wrapper provides a number of advantages, including the ability to:

- Define common functions for finding components/DOM elements:

  ```javascript
  import MyComponent from '~/path/to/my_component.vue';
  describe('MyComponent', () => {
    let wrapper;

    // this can now be reused across tests
    const findMyComponent = wrapper.findComponent(MyComponent);
    // ...
  })
  ```

- Use a `beforeEach` block to mount the component (see
  [the `createComponent` factory](#the-createcomponent-factory) for more information).
- Automatically destroy the component after the test is run with [`enableAutoDestroy`](https://v1.test-utils.vuejs.org/api/#enableautodestroy-hook)
  set in [`shared_test_setup.js`](https://gitlab.com/gitlab-org/gitlab/-/blob/d0bdc8370ef17891fd718a4578e41fef97cf065d/spec/frontend/__helpers__/shared_test_setup.js#L20).

#### Async child components

`shallowMount` will not create component stubs for [async child components](https://v2.vuejs.org/v2/guide/components-dynamic-async#Async-Components). In order to properly stub async child components, use the [`stubs`](https://v1.test-utils.vuejs.org/api/options.html#stubs) option. Make sure the async child component has a [`name`](https://v2.vuejs.org/v2/api/#name) option defined, otherwise your `wrapper`'s `findComponent` method may not work correctly.

#### The `createComponent` factory

To avoid duplicating our mounting logic, it's useful to define a `createComponent` factory function
that we can reuse in each test block. This is a closure which should reassign our `wrapper` variable
to the result of [`mount`](https://v1.test-utils.vuejs.org/api/#mount) and
[`shallowMount`](https://v1.test-utils.vuejs.org/api/#shallowMount):

```javascript
import MyComponent from '~/path/to/my_component.vue';
import { shallowMount } from '@vue/test-utils';

describe('MyComponent', () => {
  // Initiate the "global" wrapper variable. This will be used throughout our test:
  let wrapper;

  // Define our `createComponent` factory:
  function createComponent() {
    // Mount component and reassign `wrapper`:
    wrapper = shallowMount(MyComponent);
  }

  it('mounts', () => {
    createComponent();

    expect(wrapper.exists()).toBe(true);
  });

  it('`isLoading` prop defaults to `false`', () => {
    createComponent();

    expect(wrapper.props('isLoading')).toBe(false);
  });
})
```

Similarly, we could further de-duplicate our test by calling `createComponent` in a `beforeEach` block:

```javascript
import MyComponent from '~/path/to/my_component.vue';
import { shallowMount } from '@vue/test-utils';

describe('MyComponent', () => {
  // Initiate the "global" wrapper variable. This will be used throughout our test
  let wrapper;

  // define our `createComponent` factory
  function createComponent() {
    // mount component and reassign `wrapper`
    wrapper = shallowMount(MyComponent);
  }

  beforeEach(() => {
    createComponent();
  });

  it('mounts', () => {
    expect(wrapper.exists()).toBe(true);
  });

  it('`isLoading` prop defaults to `false`', () => {
    expect(wrapper.props('isLoading')).toBe(false);
  });
})
```

#### `createComponent` best practices

1. Consider using a single (or a limited number of) object arguments over many arguments.
   Defining single parameters for common data like `props` is okay,
   but keep in mind our [JavaScript style guide](javascript.md#limit-number-of-parameters) and
   stay within the parameter number limit:

   ```javascript
   // bad
   function createComponent(props, stubs, mountFn, foo) { }

   // good
   function createComponent({ props, stubs, mountFn, foo } = {}) { }

   // good
   function createComponent(props = {}, { stubs, mountFn, foo } = {}) { }
   ```

1. If you require both `mount` _and_ `shallowMount` within the same set of tests, it
   can be useful define a `mountFn` parameter for the `createComponent` factory that accepts
   the mounting function (`mount` or `shallowMount`) to be used to mount the component:

   ```javascript
   import { shallowMount } from '@vue/test-utils';

   function createComponent({ mountFn = shallowMount } = {}) { }
   ```

1. Use the `mountExtended` and `shallowMountExtended` helpers to expose `wrapper.findByTestId()`:

   ```javascript
   import { shallowMountExtended } from 'helpers/vue_test_utils_helper';
   import { SomeComponent } from 'components/some_component.vue';

   let wrapper;

   const createWrapper = () => { wrapper = shallowMountExtended(SomeComponent); };
   const someButton = () => wrapper.findByTestId('someButtonTestId');
   ```

1. Avoid using `data`, `methods`, or any other mounting option that extends component internals.

   ```javascript
   import { shallowMountExtended } from 'helpers/vue_test_utils_helper';
   import { SomeComponent } from 'components/some_component.vue';

   let wrapper;

   // bad :( - This circumvents the actual user interaction and couples the test to component internals.
   const createWrapper = ({ data }) => {
     wrapper = shallowMountExtended(SomeComponent, {
       data
     });
   };

   // good :) - Helpers like `clickShowButton` interact with the actual I/O of the component.
   const createWrapper = () => {
     wrapper = shallowMountExtended(SomeComponent);
   };
   const clickShowButton = () => {
     wrapper.findByTestId('show').trigger('click');
   }
   ```

### Setting component state

1. Avoid using [`setProps`](https://v1.test-utils.vuejs.org/api/wrapper/#setprops) to set
   component state wherever possible. Instead, set the component's
   [`propsData`](https://v1.test-utils.vuejs.org/api/options.html#propsdata) when mounting the component:

   ```javascript
   // bad
   wrapper = shallowMount(MyComponent);
   wrapper.setProps({
     myProp: 'my cool prop'
   });

   // good
   wrapper = shallowMount({ propsData: { myProp: 'my cool prop' } });
   ```

   The exception here is when you wish to test component reactivity in some way.
   For example, you may want to test the output of a component when after a particular watcher has
   executed. Using `setProps` to test such behavior is okay.

1. Avoid using [`setData`](https://v1.test-utils.vuejs.org/api/wrapper/#setdata) which sets the
   component's internal state and circumvents testing the actual I/O of the component.
   Instead, trigger events on the component's children or other side-effects to force state changes.

### Accessing component state

1. When accessing props or attributes, prefer the `wrapper.props('myProp')` syntax over
   `wrapper.props().myProp` or `wrapper.vm.myProp`:

   ```javascript
   // good
   expect(wrapper.props().myProp).toBe(true);
   expect(wrapper.attributes().myAttr).toBe(true);

   // better
   expect(wrapper.props('myProp').toBe(true);
   expect(wrapper.attributes('myAttr')).toBe(true);
   ```

1. When asserting multiple props, check the deep equality of the `props()` object with
   [`toEqual`](https://jestjs.io/docs/expect#toequalvalue):

   ```javascript
   // good
   expect(wrapper.props('propA')).toBe('valueA');
   expect(wrapper.props('propB')).toBe('valueB');
   expect(wrapper.props('propC')).toBe('valueC');

   // better
   expect(wrapper.props()).toEqual({
     propA: 'valueA',
     propB: 'valueB',
     propC: 'valueC',
   });
   ```

1. If you are only interested in some of the props, you can use
   [`toMatchObject`](https://jestjs.io/docs/expect#tomatchobjectobject). Prefer `toMatchObject`
   over [`expect.objectContaining`](https://jestjs.io/docs/expect#expectobjectcontainingobject):

   ```javascript
   // good
   expect(wrapper.props()).toEqual(expect.objectContaining({
     propA: 'valueA',
     propB: 'valueB',
   }));

   // better
   expect(wrapper.props()).toMatchObject({
     propA: 'valueA',
     propB: 'valueB',
   });
   ```

### Testing props validation

When checking component props use `assertProps` helper. Props validation failures will be thrown as errors:

```javascript
import { assertProps } from 'helpers/assert_props'

// ...

expect(() => assertProps(SomeComponent, { invalidPropValue: '1', someOtherProp: 2 })).toThrow()
```
