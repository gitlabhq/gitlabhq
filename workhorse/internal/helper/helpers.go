// Package helper provides utility functions for various tasks
package helper

import (
	"errors"
	"mime"
	"net/url"
	"os"
	"path/filepath"
	"strings"

	"gitlab.com/gitlab-org/gitlab/workhorse/internal/log"
)

// OpenFile opens a file at the specified path and returns its properties
func OpenFile(path string) (file *os.File, fi os.FileInfo, err error) {
	file, err = os.Open(filepath.Clean(path))
	if err != nil {
		return
	}

	defer func() {
		if err != nil {
			_ = file.Close()
		}
	}()

	fi, err = file.Stat()
	if err != nil {
		return
	}

	// The os.Open can also open directories
	if fi.IsDir() {
		err = &os.PathError{
			Op:   "open",
			Path: path,
			Err:  errors.New("path is directory"),
		}
		return
	}

	return
}

// URLMustParse parses the given string as a URL
func URLMustParse(s string) *url.URL {
	u, err := url.Parse(s)
	if err != nil {
		log.WithError(err).WithFields(log.Fields{"url": s}).Fatal("urlMustParse")
	}
	return u
}

// IsContentType checks if the actual content type matches the expected content type
func IsContentType(expected, actual string) bool {
	parsed, _, err := mime.ParseMediaType(actual)
	return err == nil && parsed == expected
}

// IsURL checks if the given string starts with http:// or https://
func IsURL(path string) bool {
	return strings.HasPrefix(path, "http://") || strings.HasPrefix(path, "https://")
}
