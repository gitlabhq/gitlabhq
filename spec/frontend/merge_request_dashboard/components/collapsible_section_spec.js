import { nextTick } from 'vue';
import { shallowMountExtended } from 'helpers/vue_test_utils_helper';
import LocalStorageSync from '~/vue_shared/components/local_storage_sync.vue';
import CrudComponent from '~/vue_shared/components/crud_component.vue';
import CollapsibleSection from '~/merge_request_dashboard/components/collapsible_section.vue';

describe('Merge request dashboard collapsible section', () => {
  let wrapper;

  const collapseToggle = () => wrapper.findByTestId('crud-collapse-toggle');
  const sectionContent = () => wrapper.findByTestId('crud-body');
  const emptyState = () => wrapper.findByTestId('crud-empty');
  const findLocalStorageSync = () => wrapper.findComponent(LocalStorageSync);

  function createComponent({ count = 3, hasMergeRequests = count > 0, loading = false } = {}) {
    wrapper = shallowMountExtended(CollapsibleSection, {
      slots: {
        default: 'content',
      },
      propsData: {
        id: 'list',
        title: 'Approved',
        count,
        hasMergeRequests,
        loading,
      },
      stubs: {
        CrudComponent,
      },
    });
  }

  it('renders section', () => {
    createComponent();

    // Transform snapshot for Vue2 compatiblity
    expect(wrapper.html().replace(/ison=/g, 'is-on=')).toMatchSnapshot();
  });

  it('show empty state when count is 0', () => {
    createComponent({ count: 0 });

    expect(emptyState().exists()).toBe(true);
  });

  it('hides badge when count is null', () => {
    createComponent({ count: null });

    expect(wrapper.findByTestId('merge-request-list-count').exists()).toBe(false);
  });

  it('expands collapsed content', async () => {
    createComponent({ count: 1 });

    collapseToggle().vm.$emit('click');

    await nextTick();

    expect(sectionContent().exists()).toBe(false);

    collapseToggle().vm.$emit('click');

    await nextTick();

    expect(sectionContent().exists()).toBe(true);
    expect(sectionContent().text()).toContain('content');
  });

  it('displays content when count is hidden', () => {
    createComponent({ hasMergeRequests: true, count: null });

    expect(sectionContent().exists()).toBe(true);
  });

  it('displays content when loading', () => {
    createComponent({ hasMergeRequests: false, loading: true, count: null });

    expect(sectionContent().exists()).toBe(true);
  });

  describe('collapsed state sync', () => {
    it('collapses content when local storage value is set to false', async () => {
      createComponent({ count: 1 });

      findLocalStorageSync().vm.$emit('input', false);

      await nextTick();

      expect(sectionContent().exists()).toBe(false);
    });
  });
});
