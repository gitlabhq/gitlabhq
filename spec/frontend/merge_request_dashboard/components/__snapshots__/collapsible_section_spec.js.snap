// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Merge request dashboard collapsible section renders section 1`] = `
<local-storage-sync-stub
  persist="true"
  storagekey="mr_list_list"
>
  <section
    class="crud gl-bg-subtle gl-border gl-border-section gl-mt-5 gl-rounded-lg"
    id="reference-0"
  >
    <header
      class="crud-header gl-bg-section gl-border-b gl-border-section gl-flex gl-flex-wrap gl-gap-x-5 gl-gap-y-2 gl-justify-between gl-p-4 gl-pl-5 gl-pr-10 gl-relative gl-rounded-t-lg md:gl-flex-nowrap"
    >
      <div
        class="gl-flex gl-flex-col gl-grow gl-self-center"
      >
        <h2
          class="gl-font-bold gl-gap-3 gl-inline-flex gl-items-center gl-leading-normal gl-m-0 gl-text-base"
          data-testid="crud-title"
        >
          Approved
          <gl-badge-stub
            iconsize="md"
            size="sm"
            tag="span"
            target="_self"
            variant="muted"
          >
            3
          </gl-badge-stub>
        </h2>
      </div>
      <div
        class="gl-flex gl-gap-3 gl-items-center"
        data-testid="crud-actions"
      >
        <gl-button-stub
          aria-label="Approved list help popover"
          buttontextclasses=""
          category="primary"
          class="gl-mr-2 gl-self-center"
          icon="information-o"
          size="medium"
          tag="button"
          title=""
          type="button"
          variant="link"
        />
        <div
          class="gl-absolute gl-border-l gl-border-l-section gl-h-6 gl-pl-3 gl-right-5 gl-top-4"
        >
          <gl-button-stub
            aria-controls=""
            aria-expanded="true"
            aria-label="Collapse"
            buttontextclasses=""
            category="tertiary"
            class="-gl-mr-2 btn-icon gl-self-start"
            data-testid="crud-collapse-toggle"
            icon=""
            size="small"
            tag="button"
            title="Collapse"
            type="button"
            variant="default"
          >
            <gl-animated-chevron-lg-down-up-icon-stub
              is-on="true"
              variant="default"
            />
          </gl-button-stub>
        </div>
      </div>
    </header>
    <div
      class="!gl-mx-0 crud-body gl-mb-0 gl-mx-5 gl-my-4 gl-rounded-b-lg"
      data-testid="crud-body"
    >
      content
      <div
        class="crud-pagination gl-border-t gl-border-t-section gl-flex gl-justify-center gl-p-5"
        data-testid="crud-pagination"
      />
    </div>
  </section>
</local-storage-sync-stub>
`;
